### PR TITLE
Introduce an interpreter cache.

### LINKED ISSUE #782: Interpreter discovery and pyenv don't interact well
At pex runtime, interpreter discovery looks on the PATH for any binary with a name that looks like it might be a python interpreter, and executes it to get its version (and verify that it is in fact a python interpreter). 

Unfortunately, pyenv can create a large number of such files in `~/.pyenv/shims`. E.g.,

```
> ls /Users/benjyw/.pyenv/shims/ | grep -E "python\d?(.\d)?m?$"
python
python2
python2.7
python3
python3.5
python3.5m
python3.6
python3.6m
python3.7
python3.7m
>
```
Note that these aren't symlinks, but wrapper scripts, so pex's realpath deduping doesn't help.

So, pex will execute every one of these, plus whatever other python interpreters it finds, which adds up to non trivial overhead (~1 second in the above example) on every run of the pex. Note that not all of these are actually valid python interpreters - some are shim stubs that fail on purpose.

### PR DESCRIPTION
**Pull request by jsirois**
This speeds up interpreter discovery by a factor of ~10 on cache hits.
    
Fixes #782

### CODE REVIEW

**Comment by jsirois** [comment]
N.B.: This is based off #842 and will be rebased once that is submitted. The reviewable commit here is 00cf371.

File: pex/interpreter.py [hunk 1]
```diff
@@ -290,32 +293,67 @@ def _execute(cls, binary, args=None, pythonpath=None, env=None, stdin_payload=No
...
       return cls(binary, PythonIdentity.decode(identity))
 
-    return SpawnedJob.stdout(job, result_func=create_interpreter)
+    # Part of the PythonInterpreter data are environment markers that depend on the current OS
+    # release. That data can change when the OS is upgraded but (some of) the installed interpreters
+    # remain the same. As such, include the OS in the hash structure for cached interpreters.
+    os_digest = hashlib.sha1()
+    for os_identifier in platform.release(), platform.version():
+      os_digest.update(os_identifier.encode('utf-8'))
+    os_hash = os_digest.hexdigest()
```
Comment (benjyw) [inline pex/interpreter.py]: Nice! This might save us from awkward, hard-to-find bugs later on, I would guess.

**Comment by jsirois** [comment]
After ferreting out the unrelated pyenv issues here, I'm going to break off a seperate PR to address those then circle back here and rebase.

**Comment by jsirois** [comment]
This is now ready for review. Tests in test_pex_bootstrapper should fail, but that is fixed by #860.

**Comment by jsirois** [comment]
> Pex is also not taking the approach of Pants of trying to act like Python 2 is really Python 3 

... yeah, exactly. That certainly helps avoid alot of problems.

**Pull request merged**